scalopus_examples/src/embedded_catapult_recorder.cpp

```diff
@@ -0,0 +1,114 @@
+#include <scalopus_tracing/tracing.h>
+#include <scalopus_transport/transport_loopback.h>
+
+// The catapult recorder side.
+#include <scalopus_catapult/catapult_recorder.h>
+#include <scalopus_general/general_provider.h>
+#include <scalopus_general/endpoint_manager_poll.h>
+#include <scalopus_tracing/native_trace_provider.h>
+
+void c()
+{
+  TRACE_PRETTY_FUNCTION();
+  std::this_thread::sleep_for(std::chrono::milliseconds(200));
+  std::cout << "  c" << std::endl;
+  std::this_thread::sleep_for(std::chrono::milliseconds(200));
+}
+
+void b()
+{
+  TRACE_PRETTY_FUNCTION();
+  std::cout << " b" << std::endl;
+  std::this_thread::sleep_for(std::chrono::milliseconds(200));
+  c();
+  std::this_thread::sleep_for(std::chrono::milliseconds(200));
+}
+
+void a()
+{
+  TRACE_PRETTY_FUNCTION();
+  std::cout << "a" << std::endl;
+  std::this_thread::sleep_for(std::chrono::milliseconds(200));
+  b();
+  std::this_thread::sleep_for(std::chrono::milliseconds(200));
+}
+
+int main(int /* argc */, char** argv)
+{
+  // Client side to produce the tracepoints.
+  auto factory = std::make_shared<scalopus::TransportLoopbackFactory>();
+  const auto server = factory->serve();
+  server->addEndpoint(std::make_shared<scalopus::EndpointTraceMapping>());
+  server->addEndpoint(std::make_shared<scalopus::EndpointIntrospect>());
+  server->addEndpoint(std::make_shared<scalopus::EndpointNativeTraceSender>());
+  auto endpoint_process_info = std::make_shared<scalopus::EndpointProcessInfo>();
+  endpoint_process_info->setProcessName(argv[0]);
+  server->addEndpoint(endpoint_process_info);
+
+  // Catapult recorder side.
+  auto manager = std::make_shared<scalopus::EndpointManagerPoll>(factory);
+  manager->addEndpointFactory<scalopus::EndpointTraceMapping>();
+  manager->addEndpointFactory<scalopus::EndpointProcessInfo>();
+  auto native_trace_provider = std::make_shared<scalopus::NativeTraceProvider>(manager);
+  manager->addEndpointFactory(scalopus::EndpointNativeTraceSender::name, native_trace_provider);
+
+  auto catapult_recorder = std::make_shared<scalopus::CatapultRecorder>();
+  catapult_recorder->addProvider(native_trace_provider);
+  catapult_recorder->addProvider(std::make_shared<scalopus::GeneralProvider>(manager));
+
+  auto logging_function = [](const std::string& msg) { std::cout << msg << std::endl; };
+  logging_function("Logging can be enabled in the source, uncomment the lines below this one.");
+
+  manager->connect(server->getAddress());  // Connect the manager to the server.
+  catapult_recorder->start();  // start the recorder thread.
+  catapult_recorder->startInterval();  // start recording.
+
+  TRACE_THREAD_NAME("main");
+
+  for (std::size_t i=0; i < 5; i++)
+  {
+    a();
+  }
+  // need to guarantee we propagate the tracepoints before we stop the interval.
+  std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  catapult_recorder->stopInterval();
+  catapult_recorder->setupDumpFile("/tmp/traces.json");
+
+  return 0;
+}
```

Comment:
This unit test is pretty much a copy of the [embedded_catapult_server](https://github.com/iwanders/scalopus/blob/master/scalopus_examples/src/embedded_catapult_server.cpp) example, with just the server swapped for the recorder, and some `start` & `stop` actions.